gobblin-example/src/main/java/gobblin/example/wikipedia/WikipediaExtractor.java

```diff
@@ -54,58 +64,87 @@
+  private static final DateTimeFormatter WIKIPEDIA_TIMESTAMP_FORMAT = DateTimeFormat.forPattern("YYYYMMddHHmmss");
+
+  public static final String MAX_REVISION_PER_PAGE = "gobblin.wikipediaSource.maxRevisionsPerPage";
+  public static final int DEFAULT_MAX_REVISIONS_PER_PAGE = -1;
 
-  private static final String SOURCE_PAGE_TITLES = "source.page.titles";
-  private static final String SOURCE_REVISIONS_CNT = "source.revisions.cnt";
-  private static final String WIKIPEDIA_API_ROOTURL = "wikipedia.api.rooturl";
-  private static final String WIKIPEDIA_AVRO_SCHEMA = "wikipedia.avro.schema";
+  public static final String SOURCE_PAGE_TITLES = "source.page.titles";
+  public static final String BOOTSTRAP_PERIOD = "wikipedia.source.bootstrap.lookback";
+  public static final String DEFAULT_BOOTSTRAP_PERIOD = "P2D";
+  public static final String WIKIPEDIA_API_ROOTURL = "wikipedia.api.rooturl";
+  public static final String WIKIPEDIA_AVRO_SCHEMA = "wikipedia.avro.schema";
 
   private static final String JSON_MEMBER_QUERY = "query";
   private static final String JSON_MEMBER_PAGES = "pages";
   private static final String JSON_MEMBER_REVISIONS = "revisions";
   private static final String JSON_MEMBER_PAGEID = "pageid";
   private static final String JSON_MEMBER_TITLE = "title";
 
-  private static final Splitter SPLITTER = Splitter.on(",").omitEmptyStrings().trimResults();
 
   private static final Gson GSON = new Gson();
 
-  private final WorkUnitState workUnitState;
-  private final WorkUnit workUnit;
   private final WikiResponseReader reader;
-  private final int revisionsCnt;
   private final String rootUrl;
   private final String schema;
-  private final Queue<String> requestedTitles;
-  private final int numRequestedTitles;
-  private Queue<JsonElement> recordsOfCurrentTitle;
+  private final String requestedTitle;
+  private final int batchSize;
+  private final long lastRevisionId;
+  private Queue<JsonElement> currentBatch;
+  private final ImmutableMap<String, String> baseQuery;
+  private final WorkUnitState workUnitState;
+  private final int maxRevisionsPulled;
 
   private class WikiResponseReader implements Iterator<JsonElement> {
 
+    private long latestPulledRevision;
+    private long revisionsPulled = 0;
+
+    public WikiResponseReader(long latestPulledRevision) {
+      this.latestPulledRevision = latestPulledRevision;
+    }
+
     @Override
     public boolean hasNext() {
-      if (!WikipediaExtractor.this.recordsOfCurrentTitle.isEmpty()) {
-        return true;
-      } else if (WikipediaExtractor.this.requestedTitles.isEmpty()) {
+
+      if (WikipediaExtractor.this.maxRevisionsPulled > -1
+          && this.revisionsPulled >= WikipediaExtractor.this.maxRevisionsPulled) {
+        WikipediaExtractor.this.workUnitState.setActualHighWatermark(new LongWatermark(this.latestPulledRevision));
+        LOG.info(String.format("Pulled max number of records %d, final revision pulled %d.", this.revisionsPulled,
```

Comment:
LOG.info("Pulled max number of records {}, final revision pulled {}.", this.revisionsPulled,
 this.latestPulledRevision)